### PR TITLE
Hint at what orientation means in the error message

### PR DESCRIPTION
**Pull request by asinghvi17**
Small PR that changes the error text when you pass in e.g. boundary nodes with the wrong orientation.  Feel free to close/edit if this doesn't make sense.

Should DelaunayTriangulation just reverse the order of points and go on with life?

### CODE REVIEW

**Comment by DanielVandH** [comment]
The advice depends on whether the curve is contiguous or has multiple sections.

```julia
julia> v = [[1, 2, 3], [3, 4, 5], [5, 6, 1]];

julia> rev = [[1, 6, 5], [5, 4, 3], [3, 2, 1]];

julia> reverse(v) # the sections don't line up
3-element Vector{Vector{Int64}}:
 [5, 6, 1]
 [3, 4, 5]
 [1, 2, 3]

julia> reverse(reverse.(v)) == rev # need to reverse sections and the entire curve
true

julia> v_contiguous = [1, 2, 3, 4, 5, 1]; rev_contiguous = [1, 5, 4, 3, 2, 1];

julia> reverse(v_contiguous) == rev_contiguous # good
true

julia> reverse(reverse.(v_contiguous))
ERROR: MethodError: no method matching reverse(::Int64)
The function `reverse` exists, but no method is defined for this combination of argument types.

Closest candidates are:
  reverse(::Base.AnnotatedString)
   @ Base strings\annotated.jl:318
  reverse(::SubString{<:Base.AnnotatedString})
   @ Base strings\annotated.jl:328
  reverse(::Base.Order.Ordering)
   @ Base ordering.jl:56
  ...

Stacktrace:
 [1] _broadcast_getindex_evalf
   @ .\broadcast.jl:673 [inlined]
 [2] _broadcast_getindex
   @ .\broadcast.jl:646 [inlined]
 [3] getindex
   @ .\broadcast.jl:605 [inlined]
 [4] copy
   @ .\broadcast.jl:906 [inlined]
 [5] materialize(bc::Base.Broadcast.Broadcasted{Base.Broadcast.DefaultArrayStyle{…}, Nothing, typeof(reverse), Tuple{…}})
   @ Base.Broadcast .\broadcast.jl:867
 [6] top-level scope
   @ REPL[31]:1
Some type information was truncated. Use `show(err)` to see complete types.
```

So the advice needs to be 

- If the curve is contiguous, `reverse(curve)`.
- If the curve is made of multiple sections, `reverse(reverse.(curve))`.

**Comment by DanielVandH** [comment]
We could pass whether the curve has multiple sections or not into the struct defining the error (see `has_multiple_sections`), or just breaking the advice into those two parts is probably fine..

> Should DelaunayTriangulation just reverse the order of points and go on with life?

This is an option as well but I usually prefer not to modify the user's input at all and error instead. We could continue, but then there's a mismatch between their provided boundary nodes and those stored inside the `Triangulation`. I would also need to then add support for reversing the boundary nodes for general boundaries since they don't have to be vectors. It's easier to just error I think

File: src/algorithms/triangulation/check_args.jl [hunk 1]
```diff
@@ -87,9 +87,9 @@ end
 function Base.showerror(io::IO, err::InconsistentOrientationError)
     print(io, "InconsistentOrientationError: ")
```
Comment (DanielVandH) [inline src/algorithms/triangulation/check_args.jl:88]:
```suggestion
    print(io, "InconsistentOrientationError: ")
    str = " You can fix by passing the curve as `reverse(curve)` if the nodes are contiguous, or `reverse(reverse.(curve))` if it has multiple sections."
```

File: src/algorithms/triangulation/check_args.jl [hunk 2]
```diff
@@ -87,9 +87,9 @@ end
...
     if err.should_be_positive
-        print(io, "The orientation of the boundary curve with index ", err.index, " should be positive, but it is negative.")
+        print(io, "The orientation of the boundary curve with index ", err.index, " should be positive, but it is negative.  You can fix this by passing the curve as `reverse(curve)`.")
```
Comment (DanielVandH) [inline src/algorithms/triangulation/check_args.jl:90]:
```suggestion
        print(io, "The orientation of the boundary curve with index ", err.index, " should be positive, but it is negative.", str)
```

File: src/algorithms/triangulation/check_args.jl [hunk 3]
```diff
@@ -87,9 +87,9 @@ end
...
     else
-        print(io, "The orientation of the boundary curve with index ", err.index, " should be negative, but it is positive.")
+        print(io, "The orientation of the boundary curve with index ", err.index, " should be negative, but it is positive.  You can fix this by passing the curve as `reverse(curve)`.")
```
Comment (DanielVandH) [inline src/algorithms/triangulation/check_args.jl:92]:
```suggestion
        print(io, "The orientation of the boundary curve with index ", err.index, " should be negative, but it is positive.", str)
```

**Review by DanielVandH — CHANGES_REQUESTED** [review]
What do you think of this?

**Comment by asinghvi17** [comment]
I didn't think of the impact on node order, it does make sense to simply error out then.  Will implement a `has_multiple_sections` parameter in the error then!

**Comment by DanielVandH** [comment]
> Will implement a has_multiple_sections parameter in the error then!

This also turned out to be annoying since the error isn't actually look at the boundary nodes it's looking at the `PolygonTree`. We could change it to also take the boundary nodes but that's a bit of work

**Review by DanielVandH — APPROVED** [review]

File: src/algorithms/triangulation/check_args.jl [hunk 4]
```diff
@@ -86,10 +86,11 @@ function Base.showerror(io::IO, err::InconsistentConnectionError)
 end
 function Base.showerror(io::IO, err::InconsistentOrientationError)
     print(io, "InconsistentOrientationError: ")
+    str = " You can fix by passing the curve as `reverse(curve)` if the nodes are contiguous, or `reverse(reverse.(curve))` if it has multiple sections."
```
Comment (asinghvi17) [inline src/algorithms/triangulation/check_args.jl:89]:
```suggestion
    str = " You can fix by passing the curve as `reverse(curve)` if the nodes are contiguous (i.e., a vector of indices), or `reverse(reverse.(curve))` if it has multiple sections (i.e., a vector of vectors of indices)."

Comment (asinghvi17) [reply src/algorithms/triangulation/check_args.jl:89]: If this makes sense to you then feel free to merge!

**Comment by DanielVandH** [comment]
Ahh actually this advice also doesn't really work for curve-bounded domains where `reverse` wouldn't really be defined... I will make some edits.

**Comment by DanielVandH** [comment]
See what you think now @asinghvi17. Also added some tests.

**Comment by asinghvi17** [comment]
Looks good to me - thanks!

**Comment by DanielVandH** [comment]
Thanks!

**Pull request merged**